File: src/Collega.Infrastructure/Security/Pbkdf2PasswordHasher.cs
using System.Security.Cryptography;
using System.Text;
using Collega.Application.Abstractions;

namespace Collega.Infrastructure.Security;

/// <summary>
/// PBKDF2 (RFC 2898) password hashing via the .NET BCL — deliberately not
/// Microsoft.AspNetCore.Identity or a third-party hasher, per CLAUDE.md's "no new NuGet packages
/// without approval". Stored format is self-describing and versioned
/// (<c>PBKDF2.{iterations}.{saltBase64}.{hashBase64}</c>) so a future algorithm/iteration-count
/// upgrade doesn't invalidate existing hashes.
/// </summary>
public sealed class Pbkdf2PasswordHasher : IPasswordHasher
{
    private const string Prefix = "PBKDF2";
    private const int SaltSizeBytes = 16;
    private const int HashSizeBytes = 32;
    private const int Iterations = 100_000;
    private static readonly HashAlgorithmName Algorithm = HashAlgorithmName.SHA256;

    public string Hash(string password)
    {
        var salt = RandomNumberGenerator.GetBytes(SaltSizeBytes);
        var hash = Rfc2898DeriveBytes.Pbkdf2(Encoding.UTF8.GetBytes(password), salt, Iterations, Algorithm, HashSizeBytes);

        return string.Join('.', Prefix, Iterations, Convert.ToBase64String(salt), Convert.ToBase64String(hash));
    }

    public bool Verify(string password, string passwordHash)
    {
        var parts = passwordHash.Split('.');
        if (parts.Length != 4 || parts[0] != Prefix || !int.TryParse(parts[1], out var iterations))
        {
            return false;
        }

        byte[] salt;
        byte[] expectedHash;
        try
        {
            salt = Convert.FromBase64String(parts[2]);
            expectedHash = Convert.FromBase64String(parts[3]);
        }
        catch (FormatException)
        {
            return false;
        }

        var actualHash = Rfc2898DeriveBytes.Pbkdf2(Encoding.UTF8.GetBytes(password), salt, iterations, Algorithm, expectedHash.Length);
        return CryptographicOperations.FixedTimeEquals(actualHash, expectedHash);
    }
}

File: packages/infrastructure/src/security/pbkdf2-password-hasher.ts
import { pbkdf2Sync, randomBytes, timingSafeEqual } from 'node:crypto'
import type { PasswordHasher } from '@collega/application/auth'

const PREFIX = 'PBKDF2'
const SALT_SIZE_BYTES = 16
const HASH_SIZE_BYTES = 32
const ITERATIONS = 100_000
const DIGEST = 'sha256'

/**
 * PBKDF2 (RFC 2898) password hashing via Node's `node:crypto` - ports .NET's
 * `Pbkdf2PasswordHasher` byte-for-byte, including its stored format
 * (`PBKDF2.{iterations}.{saltBase64}.{hashBase64}`), so an iteration-count upgrade later doesn't
 * invalidate existing hashes.
 */
export class Pbkdf2PasswordHasher implements PasswordHasher {
  hash(password: string): string {
    const salt = randomBytes(SALT_SIZE_BYTES)
    const hash = pbkdf2Sync(password, salt, ITERATIONS, HASH_SIZE_BYTES, DIGEST)
    return [PREFIX, ITERATIONS, salt.toString('base64'), hash.toString('base64')].join('.')
  }

  verify(password: string, passwordHash: string): boolean {
    const parts = passwordHash.split('.')
    const [prefix, iterationsText, saltBase64, hashBase64] = parts
    if (
      parts.length !== 4 ||
      prefix !== PREFIX ||
      iterationsText === undefined ||
      saltBase64 === undefined ||
      hashBase64 === undefined
    ) {
      return false
    }

    const iterations = Number(iterationsText)
    if (!Number.isInteger(iterations) || iterations <= 0) {
      return false
    }

    const salt = Buffer.from(saltBase64, 'base64')
    const expectedHash = Buffer.from(hashBase64, 'base64')
    const actualHash = pbkdf2Sync(password, salt, iterations, expectedHash.length, DIGEST)

    return actualHash.length === expectedHash.length && timingSafeEqual(actualHash, expectedHash)
  }
}
